clusterspec/clusterspec.go

```diff
@@ -0,0 +1,133 @@
+package clusterspec
+
+import (
+	"net/url"
+	"strings"
+
+	v1alpha1core "github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	v1alpha1provider "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/microerror"
+)
+
+// Config holds required configuration values for factory to be able to
+// construct provider independent Cluster configuration based on
+// ClusterGuestConfig.
+type Config struct {
+	Calico struct {
+		CIDR   int
+		MTU    int
+		Subnet string
+	}
+
+	Docker struct {
+		Daemon struct {
+			CIDR      string
+			ExtraArgs string
+		}
+	}
+
+	Etcd struct {
+		AltNames string
+		Port     int
+		Prefix   string
+	}
+
+	Kubernetes struct {
+		API struct {
+			AltNames       string
+			ClusterIPRange string
+			InsecurePort   int
+			SecurePort     int
+		}
+
+		Domain string
+
+		Hyperkube struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		IngressController struct {
+			BaseDomain string
+			Docker     struct {
+				Image string
+			}
+			InsecurePort int
+			SecurePort   int
+		}
+
+		Kubelet struct {
+			AltNames string
+			Labels   string
+			Port     int
+		}
+
+		NetworkSetup struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		SSH struct {
+			UserList []struct {
+				Name      string
+				PublicKey string
+			}
+		}
+	}
+
+	Provider struct {
+		Kind string
+	}
+}
+
+// Factory to construct provider independent cluster specifications
+type Factory interface {
```

Comment:
Single implementation interface doesn't make sense in general, but I want to enforce validation on `Config` and therefore require use of `NewFactory` constructor.

Reply:
The interfaces are usually in a `spec.go` in our code.

Reply:
Will move.